File: src/bin/0225-implement-stack-using-queues.py
# @leetup=info id=225 lang=python3 slug=implement-stack-using-queues



from typing import Optional

# @leetup=code

from collections import deque


class MyStack:
    def __init__(self):
        self.q1 = deque()
        self.q2 = deque()

    def push(self, x: int) -> None:
        self.q1.append(x)

    def pop(self) -> int:
        assert not self.empty()
        self.shorten_q1()
        return self.q1.popleft()

    def top(self) -> int:
        assert not self.empty()
        self.shorten_q1()
        return self.q1[0]

    def shorten_q1(self):
        if not self.q1:
            # q1: []
            # q2: [1, 2, 3]
            self.q1, self.q2 = self.q2, self.q1
        # q1: [1, 2, 3]
        # q2: []
        while len(self.q1) > 1:
            self.q2.append(self.q1.popleft())
        # q1: [3]
        # q2: [1, 2]

    def empty(self) -> bool:
        return (not self.q1) and (not self.q2)


# @leetup=code

File: src/bin/0225-implement-stack-using-queues.rs
// @leetup=info id=225 lang=rust slug=implement-stack-using-queues

// @leetup=code

#[allow(unused_imports)]
use std::collections::*;

struct MyStack {
    q1: VecDeque<i32>,
    q2: VecDeque<i32>,
}


/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl MyStack {

    pub fn new() -> Self {
        MyStack {
            q1: VecDeque::new(),
            q2: VecDeque::new(),
        }
    }

    pub fn push(&mut self, x: i32) {
        self.q1.push_back(x);
    }

    pub fn pop(&mut self) -> i32 {
        assert!(!self.empty());
        self.shorten_q1();
        self.q1.pop_front().unwrap()
    }

    pub fn top(&mut self) -> i32 {
        assert!(!self.empty());
        self.shorten_q1();
        *self.q1.get(0).unwrap()
    }

    fn shorten_q1(&mut self) {
        if self.q1.is_empty() {
            std::mem::swap(&mut self.q1, &mut self.q2);
        }
        while self.q1.len() > 1 {
            self.q2.push_back(self.q1.pop_front().unwrap());
        }
    }

    pub fn empty(&self) -> bool {
        self.q1.is_empty() && self.q2.is_empty()
    }
}

/**
 * Your MyStack object will be instantiated and called as such:
 * let obj = MyStack::new();
 * obj.push(x);
 * let ret_2: i32 = obj.pop();
 * let ret_3: i32 = obj.top();
 * let ret_4: bool = obj.empty();
 */
// @leetup=code

struct Solution;

fn main() {
    let _ = MyStack::new();
}
